evap/student/views.py

```diff
@@ -48,15 +53,34 @@ def index(request):
     semester_list = [dict(
         semester_name=semester.name,
         id=semester.id,
-        is_active=semester.is_active,
         results_are_archived=semester.results_are_archived,
         grade_documents_are_deleted=semester.grade_documents_are_deleted,
         evaluations=[evaluation for evaluation in evaluations if evaluation.course.semester_id == semester.id]
     ) for semester in semesters]
 
+    unfinished_evaluations = list(
+        Evaluation.objects
+        .filter(participants=request.user, state__in=['prepared', 'editor_approved', 'approved', 'in_evaluation'])
+        .exclude(voters=request.user)
+        .prefetch_related('course__responsibles', 'course__type', 'course__semester', 'participants')
```

Comment:
(both here and above)

Assuming the page just shows the participant count, it might be more appropriate to let the DB do the counting and prefetching that.

What about the voter count? that's also shown on the page, isn't it?

there are two methods both called `get_evaluations_with_prefetched_data` in the codebase, which both prefetch the voter and participant count. in case you're doing it here as well, could you look into factoring that out? (not sure how easy that would be)

Reply:
@He3lixxx if you're interested you're welcome to continue with this. otherwise i'll have a look into it.

Reply:
I won't be able to work on this until wednesday.

Reply:
> Assuming the page just shows the participant count, it might be more appropriate to let the DB do the counting and prefetching that.

In general, I agree. However, I also think that using `num_participants` is a good idea, to have a global accessor that can be easily changed when needed (For example, we might want to exclude users that dropped out of the course in the future?). So, I'd only do this if it measurably improves performance. What do you think?

On the test dataset, if I put all 950 users in as participants on one of the unfinished_evaluations (Gamification für OpenHPI), the time spent in between here and after the `render()` call varies between 0.18 and 0.23 seconds. If I make the database count the participants (`.annotate(num_participants=Count("participants", distinct=True))
`, this should "override" the cached_property, right?), I get a wrong result (1 instead of 950) and a speedup of maybe 0.01 seconds (0.176 to 0.26 seconds in ~5 tests)

> there are two methods both called get_evaluations_with_prefetched_data in the codebase, which both prefetch the voter and participant count. in case you're doing it here as well, could you look into factoring that out? (not sure how easy that would be)

Both fetch the counts in an extra query and then manually attach the results to the evaluation objects. The stated reason is:
```
        # these annotates and the zip below could be replaced by something like this, but it was 2x slower:
        # annotate(num_participants=Coalesce('_participant_count', Count("participants", distinct=True)))
```
I consider this a rather situational hack, that I'd also want to keep in as little code as possible, so I wouldn't want to use it here if we don't see similar improvements.


Voter count is not shown on the page for unfinished evaluations. For the finished evaluations, we prefetch `voters`, so the same ideas might apply there.
However, we need both prefetches in lines 49 and 50 to determine if our user has to vote or already voted, so here, the count is not sufficient:
```
        evaluation.participates_in = request.user in evaluation.participants.all()
        evaluation.voted_for = request.user in evaluation.voters.all()
```

I tried to make django compute both on the database side. With these changes, I get response times from 390 to 450ms. Without them, I get 450 to 530ms response times, however, still some of the shown results are incorrect. I pushed the code to b3e02b89. Do you see what's wrong there?

Reply:
I don't see it, but the docs say that [Combining multiple aggregations with annotate() will yield the wrong results](https://docs.djangoproject.com/en/3.1/topics/db/aggregation/#combining-multiple-aggregations) except for Counts with distinct=True. so you might need subqueries or separate queries after all.

In any case, it looks like the code did the same thing before. So i don't think it should block you from merging. Though i still think it would be better to fix this. It's not even because of the counting itself, I'm more bothered that the code fetches (for an average 6th semester bachelor student) like 2k UserProfiles from the database for no good reason.

Reply:
Okay, so I found one source of error. We had a query like this:
```
Evaluation.objects
.filter(participants=request.user, state__in=['prepared', 'editor_approved', 'approved', 'in_evaluation'])
.exclude(voters=request.user)
.annotate(num_participants=Count("participants", distinct=True))
.prefetch_related('course__responsibles', 'course__type', 'course__semester')
```

if you think of this in terms of a database operation, the `participants` m2m field is just another table that is joined in. In the SQL query, the `participants=request.user` will work by simply discarding all the rows where, after the join, the participant is not `request.user`. So, counting the participants afterwards obviously will always return "1", as for each evaluation, all other participants are discarded except for our user.

This can be fixed by simply moving the annotation up above the filter for participants. We will get a correct count then. I think something similar will work for the other broken query.

Reply:
@janno42 https://github.com/He3lixxx/EvaP/commit/b741a54ea2a44c962953c07050b1ff72111fcacb solves this discussion and passes all tests. Feel free to cherry-pick it.

Reply:
Oh, I just noticed, my commit is not taking care of `_participant_count` after archiving, so it might be broken. I don't think a student should see an evaluation with _participant_count set on their page. However, this might come back to us if for some reason this ever changes.

`get_evaluations_with_prefetched_data` solves this by using `Coalesce` to use either the `Count("participants")` or `_participant_count`. I will think a bit about how we could potentially consolidate the two `get_evaluations_with_prefetched_data` methods and this query at the student index view.

Reply:
Alright, so probably, the reason why the annotations in results/views.py:get_evaluations_with_prefetched_data were slow is explained here: https://stackoverflow.com/a/55482574/12345551
tl;dr: We create a big cartesian product and then count distinct entries. This blows quadratically in size, that's why it was slow. The post suggests a "better" fix with subqueries that allows to circumvent the manual zipping and attribute transfer.

@karyon with that, we should be able to make a `annotate_evaluations_with_participant_and_voter_counts` method and use that in th both `get_evaluations_with_prefetched_data` methods. That would be what you meant when you said "factor that out", right?

I'll try to code that.

Reply:
I added another commit (https://github.com/He3lixxx/EvaP/commit/f66cd149734ed75e7a267561913cd434cf403a82) doing what I announced there. Results, Student index and Semester index have similar performance, so it does not seem to impact performance.

Diff of both commits can be found here
https://github.com/janno42/EvaP/compare/evaluationperiod_student...He3lixxx:evaluationperiod_student_test